src/protocol/vm/adapter_contract.rs

```diff
@@ -93,17 +95,48 @@ where
         let res = self
             .call("swap", args, block, None, overwrites, None, U256::zero())
             .await?;
-        let received_amount = res.return_value[0]
-            .clone()
-            .into_uint()
-            .unwrap();
-        let gas_used = res.return_value[1]
-            .clone()
-            .into_uint()
-            .unwrap();
-        let price = self
-            .calculate_price(res.return_value[2].clone())
-            .unwrap()[0];
+
+        let (received_amount, gas_used, price) = if let Token::Tuple(return_value) =
+            res.return_value[0].clone()
+        {
+            let received_amount = match &return_value[0] {
+                Token::Uint(amount) => *amount,
+                _ => {
+                    return Err(SimulationError::DecodingError(
+                        "Expected a uint for received_amount".into(),
+                    ));
+                }
+            };
+
+            let gas_used = match &return_value[1] {
+                Token::Uint(gas) => *gas,
+                _ => {
+                    return Err(SimulationError::DecodingError(
+                        "Expected a uint for gas_used".into(),
+                    ));
+                }
+            };
+
+            let price_token = match &return_value[2] {
+                Token::Tuple(elements) => Token::Array(vec![Token::Tuple(elements.clone())]),
+                _ => {
+                    return Err(SimulationError::DecodingError(
+                        "Expected a tuple for price_token".into(),
+                    ));
+                }
+            };
+            let price = self
+                .calculate_price(price_token)?
+                .first()
+                .cloned()
+                .ok_or(SimulationError::DecodingError(
+                    "Expected at least one element in the calculated price".into(),
+                ))?;
+
+            Ok((received_amount, gas_used, price))
+        } else {
+            Err(SimulationError::DecodingError("Expected return_value to be a Token::Tuple".into()))
+        }?;
```

Comment:
This can be made easier to read/follow with pattern matching. Something like:
```suggestion
     let (received_amount, gas_used, price) = if let Token::Tuple(ref return_value) = res.return_value[0] {
        match &return_value[..] {
            [Token::Uint(amount), Token::Uint(gas), Token::Tuple(price_elements)] => {
                let received_amount = *amount;
                let gas_used = *gas;
    
                let price_token = Token::Array(vec![Token::Tuple(price_elements.clone())]);
                let price = self
                    .calculate_price(price_token)?
                    .first()
                    .cloned()
                    .ok_or_else(|| SimulationError::DecodingError(
                        "Expected at least one element in the calculated price".into(),
                    ))?;
    
                Ok((received_amount, gas_used, price))
            }
            _ => Err(SimulationError::DecodingError("Expected tuple elements with correct types".into())),
        }
    } else {
        Err(SimulationError::DecodingError("Expected return_value to be a Token::Tuple".into()))
    }?;
```

Reply:
Also can we change the error wording to not use the 'expect' pattern?